### PR TITLE
up_front_heap_allocation

### PR DESCRIPTION
**Pull request by juj**
Add a shell preamble that allocates the heap memory for asm.js/wasm page up front before any other page content, to maximize the stability for succeeding with the allocation in fragmented address space scenarios. #4712.

### CODE REVIEW

File: src/shell.html [hunk 1]
```diff
@@ -3,6 +3,37 @@
...
+      var supportsWasm = (typeof WebAssembly === 'object' && typeof WebAssembly.Memory === 'function');
+      var page_size = supportsWasm ? 64 * 1024 : 16 * 1024 * 1024;
+      var TOTAL_MEMORY = page_size * Math.ceil({{{ TOTAL_MEMORY }}} / page_size);
+      var MAX_MEMORY = Math.max(TOTAL_MEMORY, page_size * Math.ceil({{{ BINARYEN_MEM_MAX }}} / page_size));
+
+      if (supportsWasm) {
+        Module['wasmMemory'] = new WebAssembly.Memory({ initial: TOTAL_MEMORY / page_size, maximum: MAX_MEMORY / page_size });
+        Module['buffer'] = Module['wasmMemory'].buffer;
+      } else { // asm.js:
+        for(var mem = MAX_MEMORY; mem >= TOTAL_MEMORY; mem -= multiple) {
```
Comment (kripken) [inline src/shell.html]: space after `for`

**Comment by kripken** [comment]
My main concern here is that this uses total memory at compile time, but we let people change it at runtime, and this is common from what I've seen. E.g. in a pre-js. So this would be a change in html builds.

We should document that, and also add an assert that checks total memory wasn't modified (in the location where we used to allocate the memory, in preamble.js) with a clear error that explains what's going on. Possibly we want this to be an option, so that error can suggest "if you don't want this, use this option"? And we should test that.

File: src/shell.html (hunk 1 above)
Comment (juj) [inline src/shell.html]: One day I'll get this, I promise.. :)

File: src/shell.html [hunk 2]
```diff
@@ -3,6 +3,37 @@
...
+          try {
+            Module['buffer'] = new ArrayBuffer(mem);
+            if (Module['buffer'].byteLength != TOTAL_MEMORY) throw 'Out of memory';
```
Comment (juj) [inline src/shell.html]: I want to provide the same mechanism that WebAssembly gives in the sense that one can ask to allocate way more (MAX_MEMORY), but is able to settle for a minimum required amount as well (TOTAL_MEMORY). This is something that has occasionally come up as useful.

Thought it to be nice and symmetric to reuse the same variable name that currently does this for Binaryen/WebAssembly as well. (although perhaps a bit odd then that it is named after Binaryen)

Comment (juj) [inline src/shell.html]: Oh also, the check of arrayBuffer.byteLength being exactly what was asked? That comes from a bug we have seen in the wild that sometimes it is possible to allocate an ArrayBuffer, but the allocation actually fails and the byteLength = 0. That is why I check for that the allocated size is what was asked. Firefox has this kind of bug with "new ArrayBuffer(4*1024*1024*1024);", that does not fail, but returns an ArrayBuffer of size 0.

**Comment by juj** [comment]
The shift is from doing the allocation in preamble.js to doing the allocation in the shell.html file, not to compile time. Some scenarios:

 - it is possible to provide a custom --shell-file directive to fully customise the control of how the heap is allocated at runtime, so this is not fixed to compile time. (But essentially moves the allocation logic from a --pre-js file to user provided shell.html file)
 - it is true that by default code in a --pre-js file can no longer be used to allocate the heap, if one does not also add a --shell-file directive. But it is still possible to provide a custom shell.html file to omit this preallocation altogether if one wants to keep using a custom --pre-js code to set the size. Though that is the scenario we are trying to deprecate, people should not allocate the heap in a --pre-js because that is a .js file that is XHRed in after the initial page load, which means there could have been an arbitrary number of other .js file XHRs and other stuff on the page load, which may have already fragmented the process. So we'd have to document that if people want to allocate the heap in a --pre-js, they should do that XHR very first and the others only after that, which would get messy very quick. I know this is not quite what you're proposing even, but that's why I'm thinking perhaps we wouldn't even want to add an option here? (people can still do their thing with the --shell-file --pre-js combo)
 - it is also possible (at the expense of garbage collection inefficiency) to set Module['buffer'] to null in a --pre-js after setting total_memory there to use the old allocation in preamble.js, when not using a --shell-file. A bit silly, but still possible.

**Comment by kripken** [comment]
Sorry for the delay here, I've been sick.

Yeah, thanks, I see that this moves it from the JS to the HTML, not compile-time as I said. But I have the same comments above - we should clearly show an error when one tries to change it in the JS, etc.

File: src/shell.html (hunk 2 above)
Comment (kripken) [inline src/shell.html]: I'm not sure what the minimum and maximum mean in asm.js. Would they only apply in memory growth mode, like in wasm? (But then asm.js validation is disabled?)

Interesting about the byteLength zero bug, got it.

**Comment by juj** [comment]
Closing this - it was important for UE4 at one point, but they are using a custom shell that does it for them.

**Pull request closed without merging**